### PR TITLE
Run as unprivileged user

### PR DESCRIPTION
**Pull request by mindovermiles262**
Runs container as unpriviledged user

### CODE REVIEW

**Comment by ozbillwang** [comment]
Thanks for the PR. 

Let's check what's the best practices as unpriviledged user.

**Comment by mindovermiles262** [comment]
Sure. AFAIK This is best practice. It runs the container as an unpriviledged user (`UID=5280`). Let me know if/how I can help push this along 😄

**Comment by ozbillwang** [comment]
Do you have any test cases to make sure the change would not break something?

This image normally run in pipeline as tasks, and the image will be run and dropped in most cases, I don't think it is required to be unprivileged users, secondary I have no extra time to make sure all tools run with this change is no impact and also for future changes.

**Comment by mindovermiles262** [comment]
I can write a new script, `test.sh` which can be run in to verify that the following components work as expected:
* Helm/Diff/Unitest/Push
* EKSCTL/AWS CLI
* Kubectl
* Kustomize
* Kubeseal

We use this image as a the base image for a job on our Kubernetes cluster. We use an external package which sets the namespace label `pod-security.kubernetes.io/enforce: restricted`. This means that every pod in the namespace must run as an unprivileged user. I don't think there is anything special about the alpine/k8s image that requires it to run as root.

UID can be any number as long as there is not another user in the container with the same UID. I just randomly chose 5280. Changing it would have no effect as long as it's not `0` (Root user)

**Comment by mindovermiles262** [comment]
Test script was added. Please let me know if you have any additional questions or concerns.

**Comment by ozbillwang** [comment]
let me think it again. Maybe we can have different tags. 

```
alpine/k8s:1.26.1 - normal one
alpine/k8s:u1.26.1 - run as unprivileged user 
```

File: Dockerfile [hunk 1]
```diff
@@ -86,3 +76,20 @@ RUN . /envfile && echo $ARCH && \
...
+RUN helm plugin install https://github.com/databus23/helm-diff && \
+    helm plugin install https://github.com/quintush/helm-unittest && \
+    helm plugin install https://github.com/chartmuseum/helm-push && \
+    rm -rf /tmp/helm-* \
+        /home/${NONROOT_USER}/.local/share/helm/plugins/helm-push/testdata \
+        /home/${NONROOT_USER}/.cache/helm/plugins/https-github.com-chartmuseum-helm-push/testdata
+
+# Run tests
+COPY test.sh /tmp/test.sh
+RUN /tmp/test.sh
```
Comment (ozbillwang) [inline Dockerfile:95]: should not run `test.sh` in image.

test can be run on local image we built first

File: test.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  kubeseal --version > /dev/null
+}
+
+test_user_unpriviledged() {
+  # Tests that user is not root
+  echo "[*] Testing User..."
+  if [ $UID = 0 ]; then echo "[!] Running as root. Failing..."; exit 1; fi
+}
+
+test_all
```
Comment (ozbillwang) [inline test.sh:43]: any error handles?

**Comment by mindovermiles262** [comment]
Not sure why you’re so hesitant to run this as an unprivileged user. By adding an additional tag you’ll have to support an additional docker file. It is also best practice and something the k8s community is pushing towards. 

Do you have a specific use case you’re concerned about?

**Comment by mindovermiles262** [comment]
```
$ REBUILD=true ./build.sh

Found k8s latest versions: 1.26.1 1.25.6 1.24.10 1.23.16
1.26.1

[ ... ]

Installed plugin: unittest
Downloading and installing helm-push v0.10.3 ...
https://github.com/chartmuseum/helm-push/releases/download/v0.10.3/helm-push_0.10.3_linux_amd64.tar.gz
Installed plugin: cm-push
Removing intermediate container 7119c5797aa3
 ---> b79e4aac7daa
Successfully built b79e4aac7daa
Successfully tagged alpine/k8s:1.26.1
Detected Helm3+
matched
[*] Testing Helm components...
[*] Testing AWS components...
[*] Testing Kubernetes components...
[*] Testing User...
1.25.6
[ ... ]
```

If I comment out `USER ${NONROOT_USER_ID}` in the `Dockerfile`, `build.sh`fails and exits when `test.sh` fails:
```
$ REBUILD=true ./build.sh

Found k8s latest versions: 1.26.1 1.25.6 1.24.10 1.23.16
1.26.1

Installed plugin: unittest
Downloading and installing helm-push v0.10.3 ...
https://github.com/chartmuseum/helm-push/releases/download/v0.10.3/helm-push_0.10.3_linux_amd64.tar.gz
Installed plugin: cm-push
Removing intermediate container 5cead69f5813
 ---> f17189b7ae35
Successfully built f17189b7ae35
Successfully tagged alpine/k8s:1.26.1
Detected Helm3+
matched
[*] Testing Helm components...
[*] Testing AWS components...
[*] Testing Kubernetes components...
[*] Testing User...
[!] Running as root. Failing...

$ echo $?
1

**Comment by ozbillwang** [comment]
> Test script was added. Please let me know if you have any additional questions or concerns.

I am not sure I have times to maintain it in the future. 

There are several images under this Org. If I start allowing unprivileged user for this image, then someone will ask for the same to other images as well.

**Comment by mindovermiles262** [comment]
If someone else needs an unprivileged user for a different image they can make a PR. I have a specific use case for this image where the user needs to be unprivileged so I did just that - made a PR to this image.

**Comment by ozbillwang** [comment]
> @ozbillwang what your reservation here? I don't see why this type of utility container needs to run as a privileged user inside of Kubernetes.

Me too.

I use these images only for the CI/CD pipeline, and they are 'run and then terminated.' I don't really care if they are running as root or not.

Let's close this pull request now.

**Comment by mindovermiles262** [comment]
Why was this closed? There is clearly a need for this container to be run without root. This adds no additional overhead to the maintenance of this project. I _really_ do not understand your hesitation. If you don’t care if it’s root or not yet someone _does_…?

**Comment by robertcoltheart** [comment]
Can we revisit this? Most organizations now have policies for running containers as non-root and this precludes us being able to use useful tools such as this.

**Comment by ozbillwang** [comment]
If that's so formal and restricted, I would recommend to only allow private images in this type of organizations

Build own images 

the main reason is, I don't have that much time to maintain two. If I open this, others will ask for similar requirement to other images I managed under this Org.

**Pull request closed without merging**